### PR TITLE
Closure iter transformation

### PR DESCRIPTION
**Pull request by yglukhov**
This is the transformation required to support `yield` in `try`, and to support closure iters in JS. Note that from the user perspective nothing is changed still. `yield-in-try` support will follow afterwards.

### CODE REVIEW

**Comment by dom96** [comment]
oooh, and this works for the C backend too?

**Comment by dom96** [comment]
Interesting. It seems you're doing something very similar to what I attempted to do in the 'async' macro, of course doing it at this level has many advantages.

What I would like to see is tests, as many tests as possible to verify the transformation is valid.

**Comment by yglukhov** [comment]
@dom96, any specific ideas? The transformations should be covered pretty well already by all the `async`, `closure`, and `iter` categories. The change is rather low-level. I'm planning to add yield-in-try tests with my yield-in-try pr later on.

**Comment by dom96** [comment]
Oh, I thought this already enabled ``yield`` in a ``try`` statement.

**Comment by yglukhov** [comment]
@dom96, not yet, sorry.

**Comment by yglukhov** [comment]
@Araq, @zah, @dom96, the tests are green finally. Ready for review.

File: compiler/transf.nim [hunk 1]
```diff
@@ -967,20 +968,22 @@ template liftDefer(c, root) =
...
     liftDeferAux(root)
 
-proc transformBody*(module: PSym, n: PNode, prc: PSym): PNode =
-  if nfTransf in n.flags or prc.kind in {skTemplate}:
-    result = n
-  else:
-    var c = openTransf(module, "")
-    result = liftLambdas(prc, n, c.tooEarly)
-    #result = n
+proc transformBody*(ctx: PContext, n: PNode, prc: PSym): PNode =
```
Comment (Araq) [inline compiler/transf.nim]: This does not take a PContext.

File: compiler/closureiters.nim [hunk 2]
```diff
@@ -0,0 +1,632 @@
...
+#   if :tmpSlLower == 2:
+#     yield 3
+
+
+import
+  intsets, strutils, options, ast, astalgo, trees, treetab, msgs, os, options,
+  idents, renderer, types, magicsys, rodread, lowerings, tables, sequtils,
+  lambdalifting
+
+type ClosureIteratorTransformationContext = object
```
[the remark this answers is not in the document]
Comment (dom96) [reply compiler/closureiters.nim]: Man, I didn't want to say anything, but I'm glad you picked up on it. I dislike that style far more than I probably should.

File: compiler/transf.nim (hunk 1 above)
Comment (yglukhov) [reply compiler/transf.nim]: @Araq, why not? All invocations of `transformBody` pass `c.module`, so I just changed them to pass `c` :). Do you suggest I should have just added `features` to `transformBody`? I could do it...

**Comment by yglukhov** [comment]
Rebased

**Comment by yglukhov** [comment]
Superseded by #7770.

**Pull request closed without merging**